sqlmesh/core/engine_adapter/spark.py

```diff
@@ -50,37 +51,141 @@ def _use_spark_session(self) -> bool:
         return True
 
     @classmethod
-    def convert_columns_to_types_to_pyspark_schema(
+    @t.overload
+    def convert_between_spark_and_sqlglot(
         cls, columns_to_types: t.Dict[str, exp.DataType]
-    ) -> t.Optional[spark_types.StructType]:
+    ) -> spark_types.StructType:
+        ...
+
+    @classmethod
+    @t.overload
+    def convert_between_spark_and_sqlglot(
+        cls, columns_to_types: spark_types.StructType
+    ) -> t.Dict[str, exp.DataType]:
+        ...
+
+    @classmethod
+    def convert_between_spark_and_sqlglot(
+        cls, columns_to_types: t.Union[t.Dict[str, exp.DataType], spark_types.StructType]
+    ) -> t.Union[t.Dict[str, exp.DataType], spark_types.StructType]:
+        """
+        Converts either a columns_to_types -> Spark Struct or a Spark Struct -> columns_to_types
+        """
         from pyspark.sql import types as spark_types
 
-        mapping = {
-            exp.DataType.Type.BOOLEAN: spark_types.BooleanType,
-            exp.DataType.Type.TEXT: spark_types.StringType,
+        PRIMITIVE_MAPPING: t.Dict[t.Any, t.Any] = {
+            exp.DataType.Type.TINYINT: spark_types.ByteType,
+            exp.DataType.Type.SMALLINT: spark_types.ShortType,
             exp.DataType.Type.INT: spark_types.IntegerType,
             exp.DataType.Type.BIGINT: spark_types.LongType,
             exp.DataType.Type.FLOAT: spark_types.FloatType,
             exp.DataType.Type.DOUBLE: spark_types.DoubleType,
             exp.DataType.Type.DECIMAL: spark_types.DecimalType,
+            # SQLGlot currently converts these to Strings
+            exp.DataType.Type.VARCHAR: spark_types.StringType,
+            exp.DataType.Type.CHAR: spark_types.StringType,
+            exp.DataType.Type.TEXT: spark_types.StringType,
+            exp.DataType.Type.BINARY: spark_types.BinaryType,
+            exp.DataType.Type.BOOLEAN: spark_types.BooleanType,
             exp.DataType.Type.DATE: spark_types.DateType,
+            exp.DataType.Type.DATETIME: spark_types.TimestampNTZType,
+            exp.DataType.Type.TIMESTAMPLTZ: spark_types.TimestampType,
+            exp.DataType.Type.TIMESTAMPTZ: spark_types.TimestampType,
             exp.DataType.Type.TIMESTAMP: spark_types.TimestampType,
         }
 
-        try:
-            return spark_types.StructType(
-                [
-                    spark_types.StructField(col_name, mapping[col_type.this]())
-                    for col_name, col_type in columns_to_types.items()
+        COMPLEX_TYPE_MAPPING: t.Dict[t.Any, t.Any] = {
+            exp.DataType.Type.ARRAY: spark_types.ArrayType,
+            exp.DataType.Type.MAP: spark_types.MapType,
+            exp.DataType.Type.STRUCT: spark_types.StructType,
+        }
+
+        SPARK_PRIMITIVE_MAPPING = {v: k for k, v in PRIMITIVE_MAPPING.items()}
```

Comment:
is the behavior for for duplicate keys to take the last one?

Reply:
Yes